test/test_log_forwarding.py

```diff
@@ -0,0 +1,134 @@
+                    f'19775:certificate/81241b8e-c875-4a22-a30e-58003ee139ae" 0 2022-12-31T23:55:00'
+                    f'.386000Z "forward" "-" "-" "172.71.0.215:80" "204" "-" "-"'
+                ]
+                expected_output = [{
+                    'actions_executed': 'forward',
+                    'chosen_cert_arn': 'arn:aws:acm:us-east-1:122796619775:certificate/81241b8e-c87'
+                                       '5-4a22-a30e-58003ee139ae',
+                    'client': '172.71.0.215',
+                    'client:port': '172.71.0.215:36056',
+                    'client_port': 36056,
+                    'domain_name': 'gitlab.dev.singlecell.gi.ucsc.edu',
+                    'elb': 'app/azul-gitlab-alb/c051f98624e68d7e',
+                    'elb_status_code': '204',
+                    'error_reason': '-',
+                    'matched_rule_priority': '0',
+                    'received_bytes': '963',
+                    'redirect_url': '-',
+                    'request': f'POST https://gitlab.dev.singlecell.gi.ucsc.edu:443/api/v4/jobs/req'
+                               f'uest?chars={chars} HTTP/1.1',
+                    'request_creation_time': '2022-12-31T23:55:00.386000Z',
+                    'request_http_version': 'HTTP/1.1',
+                    'request_method': 'POST',
+                    'request_processing_time': '0.000',
+                    'request_uri': f'https://gitlab.dev.singlecell.gi.ucsc.edu:443/api/v4/jobs/requ'
+                                   f'est?chars={chars}',
+                    'request_uri_host': 'gitlab.dev.singlecell.gi.ucsc.edu',
+                    'request_uri_path': '/api/v4/jobs/request',
+                    'request_uri_port': 443,
+                    'request_uri_query': f'chars={chars}',
+                    'request_uri_scheme': 'https',
+                    'response_processing_time': '0.000',
+                    'sent_bytes': '229',
+                    'ssl_cipher': 'ECDHE-RSA-AES128-GCM-SHA256',
+                    'ssl_protocol': 'TLSv1.2',
+                    'target': '172.71.0.215',
+                    'target:port': '172.71.0.215:80',
+                    'target:port_list': '172.71.0.215:80',
+                    'target_group_arn': 'arn:aws:elasticloadbalancing:us-east-1:122796619775:target'
+                                        'group/azul-gitlab-http/136c2d6db59941f6',
+                    'target_port': 80,
+                    'target_processing_time': '0.002',
+                    'target_status_code': '204',
+                    'target_status_code_list': '204',
+                    'time': '2022-12-31T23:55:00.388951Z',
+                    'trace_id': 'Root=1-63b0cbd4-7d218b82786295005dbf8b6d',
+                    'type': 'https',
+                    'user_agent': 'gitlab-runner 15.6.1 (15-6-stable; go1.18.8; linux/amd64)'
+                }]
+
+                log_file_contents = '\n'.join(input).encode()
```

Comment:
Do we know for sure the logs are encoded in UTF-8 and not ASCII? If not, this needs another experiment. The user agent would be a good field to experiment with. Please document the experiment, including the `curl` invocations, so that they are reproducible.

Reply:
It appears that non-ASCII characters are escaped using the same syntax as `\`:
```
$ curl -v "https://gitlab.dev.singlecell.gi.ucsc.edu/EXPERIMENT/?unicode=$(python3 -c 'print("\u1F600")')"
...
> GET /EXPERIMENT/?unicode=ὠ0 HTTP/2
...
$ grep EXPERIMENT .cache/.fr-3PK6Z1/122796619775_elasticloadbalancing_us-east-1_app.azul-gitlab-alb.c051f98624e68d7e_20230211T0405Z_172.71.1.37_4zowzm5x.log | tr ' ' '\n'
...
"GET
https://gitlab.dev.singlecell.gi.ucsc.edu:443/EXPERIMENT/?unicode=\xE1\xBD\xA00
HTTP/2.0"
...
```

Reply:
Please repeat experiment with user agent header as I recommended. Post the entire curl output.

Reply:
Result:
```
$ AWS_PROFILE=platform-hca-dev aws s3 cp s3://edu-ucsc-gi-platform-hca-dev-logs.us-east-1/alb/access/dev/gitlab/AWSLogs/122796619775/elasticloadbalancing/us-east-1/2023/02/16/122796619775_elasticloadbalancing_us-east-1_app.azul-gitlab-alb.c051f98624e68d7e_20230216T0115Z_172.71.3.222_4164doq1.log.gz - | gunzip | grep XPERIMENT
h2 2023-02-16T01:13:03.792832Z app/azul-gitlab-alb/c051f98624e68d7e 172.71.1.200:58372 172.71.0.215:80 0.001 0.060 0.000 302 302 47 596 "GET https://gitlab.dev.singlecell.gi.ucsc.edu:443/ HTTP/2.0" "\xD0\x81XPERIMENT" ECDHE-RSA-AES128-GCM-SHA256 TLSv1.2 arn:aws:elasticloadbalancing:us-east-1:122796619775:targetgroup/azul-gitlab-http/136c2d6db59941f6 "Root=1-63ed831f-109e7a726a95c4a76f15f037" "gitlab.dev.singlecell.gi.ucsc.edu" "arn:aws:acm:us-east-1:122796619775:certificate/81241b8e-c875-4a22-a30e-58003ee139ae" 0 2023-02-16T01:13:03.730000Z "forward" "-" "-" "172.71.0.215:80" "302" "-" "-"
```

Reply:
user-agent field extracted (` | cut -d' ' -f6`): `"\xD0\x81XPERIMENT"`

Reply:
Hmm, the character I used is harder to see on GitHub than it was in my terminal. In case it isn't clear, the first letter of "ЁXPERIMENT" is `\u0401`, 'CYRILLIC CAPITAL LETTER IO'